### PR TITLE
Compare semantic equivalence of FileDescriptorSets generated by codegen-pure and protoc

### PR DESCRIPTION
**Pull request by benesch**
As discussed here: https://github.com/stepancheg/rust-protobuf/pull/505#issuecomment-657289523

The first commit adds the test itself, with some very aggressive normalization rules to work around various mismatches in codegen-pure. The next two commits fix two of the easier mismatches. Details in the individual commit messages.

### CODE REVIEW

**Comment by benesch** [comment]
I'm a little weirded out by the unit tests I needed to change in the last commit. Not sure if there was a reason that a float like `10.0` was formatted with the trailing `.0`.

File: protobuf/src/text_format/lexer/float.rs [hunk 1]
```diff
@@ -53,6 +53,6 @@ mod test {
 
     #[test]
     fn test_format_protobuf_float() {
-        assert_eq!("10.0", format_protobuf_float(10.0));
+        assert_eq!("10", format_protobuf_float(10.0));
```
Comment (stepancheg) [inline protobuf/src/text_format/lexer/float.rs:56]: This change does not look good. How Google's protobuf formats floats?

File: protobuf/src/text_format/print.rs [hunk 2]
```diff
@@ -5,12 +5,14 @@ use crate::core::Message;
...
 
-fn quote_bytes_to(bytes: &[u8], buf: &mut String) {
+#[doc(hidden)]
+pub fn quote_bytes_to(bytes: &[u8], buf: &mut String) {
     for &c in bytes {
         match c {
             b'\n' => buf.push_str(r"\n"),
             b'\r' => buf.push_str(r"\r"),
             b'\t' => buf.push_str(r"\t"),
+            b'\'' => buf.push_str("\\\'"),
```
Comment (stepancheg) [inline protobuf/src/text_format/print.rs:15]: Same here. What Google does in it's protobuf impl?

Can you please submit a separate PR with just formatting changes?

Comment (benesch) [reply protobuf/src/text_format/print.rs]: Sure, rebased out all but the test itself.

File: protobuf-codegen-identical-test/tests/diff.rs [hunk 3]
```diff
@@ -82,6 +89,97 @@ fn print_diff(dir: &Path, a: &Path, b: &Path) {
...
+
+fn normalize_descriptor_set(fds: &mut FileDescriptorSet) {
+    for desc in fds.file.iter_mut() {
+        normalize_file_descriptor(desc)
+    }
+}
+
+fn normalize_file_descriptor(desc: &mut FileDescriptorProto) {
+    if !desc.has_syntax() {
+        desc.set_syntax("proto2".into())
```
Comment (stepancheg) [inline protobuf-codegen-identical-test/tests/diff.rs:127]: Could be

```
desc.set_syntax(FileDescriptorProto::default_instance().get_syntax())
```

Up to you.

File: protobuf-codegen-identical-test/tests/diff.rs [hunk 4]
```diff
@@ -82,6 +89,97 @@ fn print_diff(dir: &Path, a: &Path, b: &Path) {
...
+    }
+    for desc in desc.message_type.iter_mut() {
```
Comment (stepancheg) [inline protobuf-codegen-identical-test/tests/diff.rs:129]:
```
for desc in &mut desc.message_type.iter_mut() {
```

File: protobuf-codegen-identical-test/tests/diff.rs [hunk 5]
```diff
@@ -82,6 +89,97 @@ fn print_diff(dir: &Path, a: &Path, b: &Path) {
...
+        normalize_descriptor(desc)
+    }
+    for desc in desc.enum_type.iter_mut() {
```
Comment (stepancheg) [inline protobuf-codegen-identical-test/tests/diff.rs:132]:
```
for desc in &mut desc.enum_type {
```

File: protobuf-codegen-identical-test/tests/diff.rs [hunk 6]
```diff
@@ -82,6 +89,97 @@ fn print_diff(dir: &Path, a: &Path, b: &Path) {
...
+    desc.options.clear();
+    for field in desc.field.iter_mut() {
+        field.options.clear();
+        // TODO: don't clear default value.
+        field.clear_default_value();
+    }
+}
+
+fn pretty_message<M: protobuf::Message>(message: &M) -> String {
+    struct FormatMessage<'a, M>(&'a M);
```
Comment (stepancheg) [inline protobuf-codegen-identical-test/tests/diff.rs:172]:
```
// TODO: pull this handy utility to protobuf crate
```

File: protobuf-codegen-identical-test/tests/diff.rs [hunk 7]
```diff
@@ -163,6 +267,22 @@ where
...
+            fs::write(
+                format!("{}/{}.descriptors", pure_dir, proto_name),
+                pretty_message(&pure_descriptors),
+            )
+            .unwrap();
+            print_diff(
+                temp_dir.path(),
+                Path::new(&format!("protoc/{}.descriptors", proto_name)),
+                Path::new(&format!("pure/{}.descriptors", proto_name)),
+            );
```
Comment (stepancheg) [inline protobuf-codegen-identical-test/tests/diff.rs:285]: If you can, split this function so it could fit on one page.

File: protobuf-codegen-identical-test/tests/diff.rs [hunk 8]
```diff
@@ -82,6 +89,97 @@ fn print_diff(dir: &Path, a: &Path, b: &Path) {
...
+        .out(temp_file.path())
+        .includes(includes)
+        .inputs(inputs)
+        .write_descriptor_set()
+        .unwrap();
+    protobuf::parse_from_reader(&mut temp_file).unwrap()
+}
+
+fn pure_descriptor_set(includes: &[PathBuf], inputs: &[PathBuf]) -> FileDescriptorSet {
+    let mut codegen = protobuf_codegen_pure::parse_and_typecheck(includes, inputs).unwrap();
```
Comment (stepancheg) [inline protobuf-codegen-identical-test/tests/diff.rs:105]:
```
// TODO: expose this utility from protobuf-codege-pure crate
```

File: protobuf-codegen-identical-test/tests/diff.rs (hunk 7 above)
Comment (benesch) [reply protobuf-codegen-identical-test/tests/diff.rs:287]: I didn't see a great way of splitting this up without passing a whole bunch of parameters around. Happy to try harder if you like.

File: protobuf-codegen-identical-test/tests/diff.rs (hunk 3 above)
Comment (benesch) [reply protobuf-codegen-identical-test/tests/diff.rs:128]: Actually, that seems to return `""`, so leaving as is.

**Comment by stepancheg** [comment]
Thanks!
